lib/puppet/ssl/state_machine.rb

```diff
@@ -349,6 +362,34 @@ def next_state
     end
   end
 
+  # Class to renew a client/host certificate automatically.
+  #
+  class NeedRenewedCert < KeySSLState
+    def next_state
+      Puppet.debug(_("Renewing client certificate"))
+
+      route = @machine.session.route_to(:ca, ssl_context: @ssl_context)
+      cert = OpenSSL::X509::Certificate.new(
+        route.post_certificate_renewal(@ssl_context)[1]
+      )
+
+      # verify client cert before saving
+      next_ctx = @ssl_provider.create_context(
+        cacerts: @ssl_context.cacerts, crls: @ssl_context.crls, private_key: @private_key, client_cert: cert
+      )
+      @cert_provider.save_client_cert(Puppet[:certname], cert)
+
```

Comment:
Could we add an `info` level message that the client cert was refreshed (like we do for CA & CRL), something like: 

```ruby
      Puppet.info("Refreshed client certificate: #{@machine.digest_as_hex(cert.to_pem)}, not before '#{cert.not_before}', not after '#{cert.not_after}'")
```